pkg/macaroon/apikey.go

```diff
@@ -0,0 +1,185 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package macaroon
+
+import (
+	"bytes"
+
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+)
+
+var (
+	// Error is a general API Key error
+	Error = errs.Class("api key error")
+	// ErrFormat means that the structural formatting of the API Key is invalid
+	ErrFormat = errs.Class("api key format error")
+	// ErrInvalid means that the API Key is improperly signed
+	ErrInvalid = errs.Class("api key invalid error")
+	// ErrUnauthorized means that the API key does not grant the requested permission
+	ErrUnauthorized = errs.Class("api key unauthorized error")
+	// ErrRevoked means the API key has been revoked
+	ErrRevoked = errs.Class("api key revocation error")
+)
+
+// APIKey implements a Macaroon-backed Storj-v3 API key.
+type APIKey struct {
+	mac *Macaroon
+}
+
+// ParseAPIKey parses a given api key string and returns an APIKey if the
+// APIKey was correctly formatted. It does not validate the key.
+func ParseAPIKey(key string) (*APIKey, error) {
+	data, version, err := base58.CheckDecode(key)
+	if err != nil || version != 0 {
+		return nil, ErrFormat.New("invalid api key format")
+	}
+	mac, err := ParseMacaroon(data)
+	if err != nil {
+		return nil, ErrFormat.Wrap(err)
+	}
+	return &APIKey{mac: mac}, nil
+}
+
+// NewAPIKey generates a brand new unrestricted API key given the provided
+// server project secret
+func NewAPIKey(secret []byte) (*APIKey, error) {
+	mac, err := NewUnrestricted(secret)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+	return &APIKey{mac: mac}, nil
+}
+
+// Check makes sure that the key authorizes the provided action given the root
+// project secret and any possible revocations, returning an error if the action
+// is not authorized. 'revoked' is a list of revoked heads.
+func (a *APIKey) Check(secret []byte, action Action, revoked [][]byte) error {
+	if !a.mac.Validate(secret) {
+		return ErrInvalid.New("macaroon unauthorized")
+	}
+
+	caveats := a.mac.Caveats()
+	if len(caveats) == 0 {
+		// make sure we always check at least one empty caveat
+		caveats = [][]byte{[]byte("")}
+	}
```

Comment:
This is non-obvious, whereas ensuring the action has a time directly is obvious.

Reply:
fair